RFD 15: 2FA device management

awly opened the pull request:
Support for multiple 2FA devices (both OTP and U2F) per user, and allowing 2FA to be optional.

Required for RFD 14 and https://github.com/gravitational/teleport/issues/1929

Commit:
RFD 15: 2FA device management

klizhentas commented on rfd/0015-2fa-management.md:
```diff
@@ -0,0 +1,206 @@
+$ tsh 2fa ls
+2FA deivice name   Type   Last used
+----------------   ----   -------------------------------
+android OTP        OTP    Tue 15 Dec 2020 01:29:42 PM PST
```
shall we remove by ID instead of the name to avoid name collisions and other possible problems?

awly replied:
IDs are not very user-friendly.
We will prevent users from creating devices with existing names.
All these names are namespaced under the user anyway (see the storage section below), so there's no conflicts between users.

klizhentas replied:
We can follow github strategy where they can type first 5 letters and copy them. I just think that having user-input data as a key is dangerous for so many reasons :)

klizhentas replied:
You'd probably can still delete by ID, but let users delete by name.

awly replied:
It wouldn't be a key in backend storage path, since keys get stored in the user object.
I don't want to add IDs on top of this unless necessary, to keep things simple.

klizhentas replied:
how would you trace the key usage in the certificate? could it be public key fingerprint used as an ID? I still think that using user-input data will be a problem

awly replied:
I don't think we need to specify which specific key was used in the certificate.
Let me sketch out the cert content changes.

klizhentas replied:
This could be helpful for audit purposes, because we will be able to connect action to a specific device. We will gain ability to block the activity by device that got lost without blocking the user.

awly replied:
I get a YAGNI smell from these key IDs, but adding regardless. It's only a small amount of additional complexity.

Added cert extensions and key ID in https://github.com/gravitational/teleport/pull/5125/commits/d26feeb64c69e43c486bd7c5783b411e2e4f95c6

klizhentas commented on rfd/0015-2fa-management.md:
```diff
@@ -0,0 +1,206 @@
+Tap your *new* security key... <tap>
+2FA device "solokey" added.
+
+$ tsh 2fa ls
```
typing `tsh 2fa` seems a bit unusual due to `2` being first, nitpicking on the name, is there a better one easier to type, or is it just me.

awly replied:
How about `mfa`?

klizhentas commented on rfd/0015-2fa-management.md:
```diff
@@ -0,0 +1,206 @@
+
+### configuration
+
+The current 2FA configuration in Teleport only applies to local users and is
```
The biggest challenge is to let SSO users enroll their keys. These users are ephemeral, so if we connect any data under the username, it will get deleted. I haven't figured out how to combine those two flows, but this feature will be used by Enterprise users.

awly replied:
I remember asking this in the past, but don't remember the answer: can we simply stop expiring SSO users from our backend?

klizhentas replied:
If we do this by default, this could lead to confusion, because administrators would need to start clean those entries up periodically. 

If someone enrolls a U2F key, we can remove TTL from this SSO user and make sure it won't be set. So we will treat those users differently.

awly replied:
Are Teleport-stored users exposed anywhere admin-facing? AFAIK it's just about backend storage consumption, right?

A long TTL, like 1-2 years, could be a good balance between accumulating backend garbage and not deleting users after prolonged inactivity. Worst case: someone doesn't touch teleport for a year and after a fresh SSO login they need to re-add U2F keys.

klizhentas replied:
they are visible in the same UI alongside the local users.

awly replied:
we also store MFA data separately from other user data
we could persist MFA data (with very long TTL) and keep other user data ephemeral as it is today

Commit:
Update author email address

klizhentas commented on rfd/0015-2fa-management.md:
```diff
@@ -0,0 +1,206 @@
+New fields added under `auth_service.authentication`:
+- `second_factor_optional` - bool, defaults to `false`; when `true`, users can
+  register and login without 2FA checks, and are able to add 2FA voluntarily
+- `second_factor` - keep existing values `otp`, `u2f`, `off`, and add a new
```
I think `any` is great, can we remove extra knob `second_factor_optional`?

awly replied:
`any` here doesn't include `off` - it just means that any 2FA method is supported
We could instead have other values:
- `second_factor: optional` - allow users to enroll in 2FA, but don't enforce it
- `second_factor: on` - require all users to have at least one 2FA method enabled

does that sound better?

benarent commented on rfd/0015-2fa-management.md:
```diff
@@ -0,0 +1,206 @@
+
+Web UI management of 2FA devices should be logically similar to the CLI:
+- a page to see all enrolled devices
+- a button to enroll a new device
```
Are we planning on using WebAuthn? It has a range of built in UX for registering tokens per supported browser https://caniuse.com/?search=webauthn

AWS recently added support and it has a few good points of reference. https://aws.amazon.com/blogs/aws/multi-factor-authentication-with-webauthn-for-aws-sso/

awly replied:
U2F UX in the browser is basically identical to WebAuthn - you get a popup prompting for a tap.
It's not that clunky version where you have a textbox and the U2F token generates a bunch of random key presses.

But yes, at some point we can add WebAuthn support too.

benarent replied:
Thanks for the clarification, I guess It'll be same as our current U2F setup.. which could be improved but works.

awly replied:
What are the issues with our current U2F setup (other than being browser-only)?
Maybe I can address those along the way.

benarent replied:
The setup is currently a bit clunky, but once setup it's pretty smooth. I'll set it up again and find the rough edges.

gclawes replied:
WebAuthn is a part of FIDO2, the successor to U2F.  I believe it is more compatible on mobile as well (I noticed mobile NFC U2F doesn't currently work this week, at least on Android/Chrome).

awly replied:
AFAIK, the user interaction would be basically the same for U2F and WebAuthn.

The only difference is token support when using a browser (e.g. SSH via the web UI).
On the CLI (`kubectl` or `tsh ssh`), we'll only support USB devices for now.

benarent commented on rfd/0015-2fa-management.md:
```diff
@@ -0,0 +1,206 @@
+yubikey            U2F    Wed 16 Dec 2020 02:00:13 PM PST
+
+$ tsh 2fa add
+Adding a new 2FA device.
```
Would it be simpler and more widely supported if we off loaded this to the UI?

```
$ tsh 2fa add

Registering new 2FA Device. If browser window does not open automatically, open it by clicking on the link:
 http://127.0.0.1:53414/d2ac3d35-2cee-4229-ac4a-1ba0ae71f11c
```

Commit:
Address review feedback

- change `tsh 2fa` to `tsh mfa`
- change the auth_server config to not add any new fields

Commit:
Add UUIDs to stored keys

Commit:
Add audit log section

klizhentas commented on rfd/0015-2fa-management.md:
```diff
@@ -0,0 +1,225 @@
+  for all local users
+- `optional` (new) - users can enroll both OTP and U2F devices, and 2FA is
+  required only for users with 2FA enrolled
+- `session_only` (new) - users can enroll both OTP and U2F devices, and 2FA is
```
why do we need `session_only`? can this option always available unless it's `off` or `otp`?

awly replied:
Good point. If there are 2FA devices enrolled, users have to use them on login even if it's also required for individual sessions.

awly commented:
TODO(me): add option to validate the key attestation cert against a fixed set of CAs.
See https://developers.yubico.com/U2F/Attestation_and_Metadata/

klizhentas commented:
@zapman449

Commit:
Remove a redundant config option and add device attestation check

ikkisoft commented on rfd/0015-2fa-management.md:
```diff
@@ -0,0 +1,253 @@
+- `off` (existing) - no 2FA can be enrolled
+- `otp` (existing) - only OTP can be enrolled and is required for all local
+  users
+- `u2f` (existing) - only U2F can be enrolled and is required for all local
```
Enforcement for U2F only is particularly important as I would expect many customers to use this. In practice, U2F is the only protection agains phishing.

ikkisoft commented on rfd/0015-2fa-management.md:
```diff
@@ -0,0 +1,253 @@
+$ tsh mfa ls
+MFA deivice name   Type   ID                                     Last used
+----------------   ----   ------------------------------------   -------------------------------
+android OTP        OTP    fa004bf4-acc7-435d-8965-5f5a0a4552e8   Tue 15 Dec 2020 01:29:42 PM PST
```
I've done a bit of UX research and most SaaS apps display the **Added Date** instead of the last time of use. Google shows _Added Date_ and _Last Used_. I believe that this is done to identify malicious devices with similar names.

awly replied:
Thanks for the suggestion, added the `Added at` field.

ikkisoft commented on rfd/0015-2fa-management.md:
```diff
@@ -0,0 +1,253 @@
+Adding a new MFA device.
+Choose device type (1 - OTP, 2 - U2F): 2
+Enter device name: solokey
+Tap any *registered* security key... <tap>
```
As discussed in person, I believe that this requirement can be relaxed when the system does not perform post-login operations that would require MFA. Considering https://github.com/gravitational/teleport/pull/5125, it is however mandatory to enforce 2FA step up for add/delete operations. The usability drawback is that users that lost / broke a device might have to ask the admin for a reset.

awly commented:
@klizhentas @russjones @benarent ping for review

a-palchikov commented on rfd/0015-2fa-management.md:
```diff
@@ -155,10 +155,38 @@ New values for `auth_service.authentication.second_factor` for this:
+
+Note that existing enrolled keys will remain, even if they aren't Yubikeys.
+This is to avoid storing an attestation cert of every key in our backend, which
+we would need in order to re-check existing keys when a new CA gets added.
```
Sorry it reads a bit confusing to me: `would need in order ...` means `storing the attestation certificates in the backend` or that `every key is necessary` to recheck existing keys? I understand this is probably the latter but I'd rephrase:
```
This is to avoid storing an attestation cert of every key in our backend when
we need to re-check existing keys whenever a new CA is added.
```

awly replied:
Rephrased, hopefully it's clearer

benarent commented on rfd/0015-2fa-management.md:
```diff
@@ -0,0 +1,254 @@
+
+$ tsh mfa add
+Adding a new MFA device.
+Choose device type (1 - OTP, 2 - U2F): 2
```
We could implement Show TOTP secret in register UI / CLI #3472 to make this mfa UX a pure CLI experience for OTP tokens.

awly replied:
Added OTP enrollment via CLI example.
I'll implement this last though, potentially in 6.1, due to the approaching 6.0 deadline.

benarent commented on rfd/0015-2fa-management.md:
```diff
@@ -0,0 +1,254 @@
+$ tsh mfa ls
+MFA deivice name   Type   ID                                     Added at                          Last used
+----------------   ----   ------------------------------------   -------------------------------   -------------------------------
+android OTP        OTP    fa004bf4-acc7-435d-8965-5f5a0a4552e8   Tue 08 Dec 2020 01:29:42 PM PST   Tue 15 Dec 2020 01:29:42 PM PST
```
Do we need ID? I would suggest we only show it in debug mode if it's an internal ID.

ikkisoft replied:
Agree. Don't see much value on having it.

awly replied:
@klizhentas wanted to have IDs in addition to plain names, due to theoretical naming collisions.
I'll just remove the IDs from `tsh` output, but keep them in internal storage representation.

benarent commented on rfd/0015-2fa-management.md:
````diff
@@ -0,0 +1,254 @@
+Can't remove the only remaining MFA device.
+Please add a replacement MFA device first using "tsh mfa add".
+```
+
````
I was looping back on older issues, there is this use case for reseting/regenerating a users OTP without having to re-create the user. https://github.com/gravitational/teleport/issues/2280  It's an edge case, and maybe less important now users can setup more than one token.

awly replied:
Yeah, I think registering a new OTP device and removing the old one works.
A user will need to have access to one of the registered OTP or U2F devices though, otherwise there's a risk of 2FA bypass.

Commit:
Address review feedback

klizhentas commented on rfd/0015-2fa-management.md:
```diff
@@ -0,0 +1,269 @@
+auth_service:
+  authentication:
+    second_factor: "on" # or "u2f" or "optional"
+    u2f_device_attestation_cas:
```
@awly are we going to support those as resources via API as well as config parameters? what's the data model?

awly replied:
This is stored in the `AuthPreference` of the auth server: https://github.com/gravitational/teleport/blob/4b5bb54bf4390bdcc7c8b45424a248998c42d7e4/api/types/authentication.go#L267-L290 (along with the other U2F parameters).
The actual "storage" is the `teleport.yaml` of the auth server. There are REST endpoints for `AuthPreference` https://github.com/gravitational/teleport/blob/4b5bb54bf4390bdcc7c8b45424a248998c42d7e4/lib/auth/apiserver.go#L201-L202. It seems to be read-only in practice (only tests call `SetAuthPreference`).

`u2f_device_attestations_cas` are not in the same boat as our host/user/jwt CAs. These U2F CAs are just the certificates (not keys), usually from 3rd parties like Yubico.

klizhentas replied:
@awly can you modify the design to be API friendly? In the current form it will be impossible to use it with cloud or terraform provider using API. I can imagine for example to support paths and content as optional parameter, so folks can set up auth preference resource using API and supply CAs

awly replied:
Added an example of passing the raw CA PEM blob in the config.
It can also be passed via `SetAuthPreference` remotely.
Is this what you mean?

Commit:
Allow u2f attestation CAs as inline PEM blocks

benarent commented on rfd/0015-2fa-management.md:
````diff
@@ -0,0 +1,290 @@
+}
+```
+
+#### migration
````
```suggestion
#### Migration
```

benarent commented on rfd/0015-2fa-management.md:
```diff
@@ -0,0 +1,290 @@
+The above `LocalAuthSecrets` will be migrated by Teleport at startup in v6.
+In v7, we will remove the deprecated MFA fields from `LocalAuthSecrets.`
+
+### audit log
```
```suggestion
### Audit log
```

benarent commented on rfd/0015-2fa-management.md:
```diff
@@ -0,0 +1,290 @@
+would impose extra storage costs, add complexity and create the risk of user
+lockout.
+
+### backend storage
```
```suggestion
### Backend storage
```

benarent commented on rfd/0015-2fa-management.md:
```diff
@@ -0,0 +1,290 @@
+- `optional` (new) - users can enroll both OTP and U2F devices, and 2FA is
+  required only for users with 2FA enrolled
+
+#### restricted device vendors
```
```suggestion
#### Restricted device vendors
```

benarent commented on rfd/0015-2fa-management.md:
````diff
@@ -0,0 +1,290 @@
+Please add a replacement MFA device first using "tsh mfa add".
+```
+
+#### web UI
````
```suggestion
#### Web UI
``

benarent commented on rfd/0015-2fa-management.md:
```diff
@@ -0,0 +1,290 @@
+Web UI details, wireframes and implementation will be added later, when we have
+the capacity to do it. Initially, 2FA management is CLI-only.
+
+#### bootstrap
```
```suggestion
#### Bootstrap
```

benarent commented on rfd/0015-2fa-management.md:
```diff
@@ -0,0 +1,290 @@
+can then add 2FA devices as described above. If an existing user has at least
+one 2FA device registered, it's required during login.
+
+### configuration
```
```suggestion
### Configuration
```

benarent reviewed (approved):
My comments are addressed. A small nit, some titles are capitalised some aren't. Added suggestion to keep things the same.

Commit:
Apply suggestions from code review

Co-authored-by: Ben Arent <ben@gravitational.com>

Commit:
Merge branch 'master' into rfd/0015-2fa-management

Commit:
Merge branch 'master' into rfd/0015-2fa-management
